src/infra/core/src/utils/datasets_filtering.rs

```diff
@@ -41,10 +63,145 @@ pub fn filter_datasets_by_pattern(
                 future::ready(
                     dataset_ref_patterns
                         .iter()
-                        .any(|dataset_ref_pattern| dataset_ref_pattern.is_match(dataset_handle)),
+                        .any(|dataset_ref_pattern| dataset_ref_pattern.matches(dataset_handle)),
                 )
             })
             .map_err(Into::into)
             .boxed()
     }
 }
+
+////////////////////////////////////////////////////////////////////////////////
+
+pub fn filter_datasets_by_any_pattern(
+    dataset_repo: &dyn DatasetRepository,
+    search_svc: Arc<dyn SearchService>,
+    dataset_ref_any_patterns: Vec<DatasetRefAnyPattern>,
+) -> FilteredDatasetRefAnyStream<'_> {
+    let in_multitenant_mode = dataset_repo.is_multi_tenant();
+
+    let (all_ref_patterns, static_refs): (Vec<_>, Vec<_>) = dataset_ref_any_patterns
+        .into_iter()
+        .partition(DatasetRefAnyPattern::is_pattern);
+
+    let (remote_ref_patterns, local_ref_patterns): (Vec<_>, Vec<_>) = all_ref_patterns
+        .into_iter()
+        .partition(|pattern| pattern.is_remote_pattern(in_multitenant_mode));
+
+    let static_datasets_stream = get_static_datasets_stream(static_refs);
+    let remote_patterns_stream =
+        get_remote_datasets_stream(search_svc, remote_ref_patterns, in_multitenant_mode);
+    let local_patterns_stream = get_local_datasets_stream(dataset_repo, local_ref_patterns);
+
+    static_datasets_stream
+        .chain(remote_patterns_stream)
+        .chain(local_patterns_stream)
+        .boxed()
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+fn get_static_datasets_stream(
+    static_refs: Vec<DatasetRefAnyPattern>,
+) -> FilteredDatasetRefAnyStream<'static> {
```

Comment:
Since this function is private to the module and the stream you create here is later boxed in `filter_datasets_by_any_pattern` after chaining anyway, I think boxing here is not necessary. 

You might be able to return `impl TryStream<..>` and avoid allocation.

Same for `get_remote_datasets_stream`.

Reply:
Corrected. 3 impls that get chained and boxed at the very end